### PR TITLE
[jsscripting] Reimplement timer creation method of `ScriptExecution`

### PR DESCRIPTION
**Pull request by florian-h05**
## Description

This PR reimplements the openHAB Core API for creating timers to make this also threadsafe.

As discussed on the forum https://community.openhab.org/t/js-scripting-why-are-we-deprecated-createtimer/140748, the `setTimeout` and `setInterval` methods do not provide the functionality that advanced users need, and to avoid that those users and users coming from Rules DSL leave JS Scripting because of this missing API, we have to provide thread-safe access to it.

Furthermore,  `JSScriptServiceUtil` was introduced to provide easy access from to script services provided by core. This replaces the usage of the `JSScriptServiceUtil` from the Rules DSL parts of core and therefore reduces dependency on a part that might become an addon in the future.

openhab-js is also upgraded to 2.1.1.

## Testing

Use the following file-based script to test:

```javascript
const { time} = require('openhab');

// Note that all timers are scheduled to nearly the same time, so without the synchronization mechanism in place, this would lead to a multi-thread exception
ThreadsafeTimers.createTimer(time.toZDT(), () => {
    console.info("Hello world from createTimer without identifier")
});

ThreadsafeTimers.createTimer('createTimer', time.toZDT(), () => {
    console.info("Hello world from createTimer with identifier")
});
```

### CODE REVIEW

**Comment by florian-h05** [comment]
@rkoshak 
PR is now open.

@jlaur @jpg0 Could you please give your feedback on this?

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java
Comment (digitaldan) [inline bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java:146]: Should we remove the future from the map here ?  so `idSchedulerMapping.remove(id)` ?  I noticed the only time we remove references to futures is when we completely clear the map at init/reload time?

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java
Comment (jpg0) [inline bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java:97]: I have two questions about this method. 

1) Presumably your test should actually be:

```
ThreadsafeTimers.createTimerWithArgument(time.toZDT(), arg1, (newArg) => {
    console.info(`Hello world from createTimerWithArgument without identifier and argument: ${newArg}`);
});
```

2) What is the use case here? Why not allow the closure to capture the state? This is effectively what your original test actually did, as the passed parameter was unused. Which is why I'm not sure I understand the use-case.

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java
Comment (jpg0) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java:146]: It's also cleared on `clearTimeout`, but I agree that it should be cleared here too.

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/GraalJSScriptServiceUtil.java
Comment (J-N-K) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/GraalJSScriptServiceUtil.java:32]: I would suggest to leave out Graal. 3.4 will likely be the last one with integrated Nashorn JS, so my guess would be that someone in a later release will either install Nashorn JS or this one.

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java
Comment (florian-h05) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java:146]: I agree, I'll change this.

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java
Comment (florian-h05) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java:97]: > Presumably your test should actually be:

Yes, thanks.

> What is the use case here? Why not allow the closure to capture the state?

TBH, I am not sure what the use case is here. I do not use this in any of my scripts.

Comment (jpg0) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java:97]: In that case I would suggest that we remove the `createTimeWithArgument` methods. I _suspect_ that the reason they are here is because:

- The original Java class included this method because it's not so simple to include state when calling from Java, so passing state in (whilst providing the same callback interface) is very convenient
- These were by default just extended to scripts

I don't really see the need in JS because people will pass closures that capture the state implicitly. Having these extra methods just mean extra code to maintain, plus a more confusing API for script writers.

Comment (florian-h05) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java:97]: Good idea, I also don't see the need. 

> plus a more confusing API for script writers.

That is a really good argumen.

**Comment by florian-h05** [comment]
@jpg0 I addressed your review, you could please re-review?

**Review by digitaldan — APPROVED** [review]
LGTM

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java
Comment (jpg0) [inline bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java:33]: This is a slightly odd use of OSGi. You have a component which is activated but then permanently stores (as static) the passed in references. Typically these references would not be static, however it means that consumers of this class would need it injecting too, rather that using static references to it.

I'm not sure it (a) this is a common pattern in openHAB (in which case go ahead with it), or (b) if it will cause any real problems or not. There _may_ be a problem with plugin reloading here, as the component will be reloaded and the fields will then be reset to new values, and any consumers (via the static ref) will see updated values, whereas consumers via injection will see two versions of `JSScriptServiceUtil` where the old one internally gets updated to the values of the new one. I am also not sure if there could be issues with components hanging on to static class references (of `JSScriptServiceUtil`) and hence preventing class collection by the GC.

The alternative to all this would be to (a) make these references non-static, (b) make `ThreadsafeTimers` a component and inject this class as a `@Reference` into it, and (c) fetch an instance of `ThreadsafeTimers` from the bundle context (rather than creating your own).

Personally I think this is a better way to go, however if this is an existing pattern in openHAB then I don't have an issue with it.

Comment (jpg0) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java:33]: Oh, and doing without static refs means that you can get rid of the `@Nullable` annotations as the references can never be null.

**Review by jpg0 — APPROVED** [review]
Only issue I have is with the use of OSGi. If it's a common pattern in openHAB I'm not concerned about it though.

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java
Comment (J-N-K) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java:33]: I agree with @jpg0. This is an ugly workaround that is used in the DSL Script engine because it would (at least) require very heavy refactoring to make the OSGi components available to XText. It should not be used if there is a better way.

Comment (florian-h05) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java:33]: Could I also make `JSScriptServiceUtil` non-static, initialize it from somewhere (OpenHABGraalJSScriptEngine) and use the constructor to get `Scheduler` and `ScriptExecution` from OSGi?

Comment (florian-h05) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: @jpg0 @J-N-K See commit https://github.com/openhab/openhab-addons/pull/13695/commits/fedbbe61fbbfec6b90c0e58692f1d40b06d31dbd for how I am retrieving the OSGi components.

Comment (jpg0) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: Your updated method is slightly better, but why not just inject the dependencies? The reason for injection is to break the coupling between classes, to allow things like testing them independently. Putting all the lookup code in the constructor now ties the class to OSGi too, as well as making it more complex. Are you doing that to make the lookup from JS easier? (There is an OSGi lookup ability in JS already.)

Comment (florian-h05) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: > why not just inject the dependencies?

You are referring to the solution you propose above, correctly?

> The alternative to all this would be to (a) make these references non-static, (b) make ThreadsafeTimers a component and inject this class as a @Reference into it, and (c) fetch an instance of ThreadsafeTimers from the bundle context (rather than creating your own).

So my question is, how to get the lock object into ThreadsafeTimers when I fetch it from the bundle context.

Comment (J-N-K) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: You could inject the `ScriptExecution` to the `GraalJSScriptEngineFactory` which is already a component and then pass it in the constructor to the `OpenhabGraalJSScriptEngine`. 

A "better" but more complex but better solution would be to make a factory component that creates a thread-safe component instance for each caller. The factory component could then get the `ScriptExecution` and also the `Scheduler` via injection and create a new lock object each time it creates a new instance.

Comment (florian-h05) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: > You could inject the ScriptExecution to the GraalJSScriptEngineFactory which is already a component and then pass it in the constructor to the OpenhabGraalJSScriptEngine.

Can I also inject the `JSScriptServiceUtil` (as long as it is a `@Component`) to the `GraalJSScriptEngineFactory`?

> A "better" but more complex but better solution would be to make a factory component that creates a thread-safe component instance for each caller. The factory component could then get the ScriptExecution and also the Scheduler via injection and create a new lock object each time it creates a new instance.

I am not sure if I able to implement this. I will try.

Comment (J-N-K) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: If you need assistance, feel free to contact me by private mail.

Comment (florian-h05) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: > If you need assistance, feel free to contact me by private mail.

Thanks for the offer.

Could you please check if my current implementation is okay?

Regarding the "better" solution: I am asking myself if is worth the additional work.
If we need access to more OSGi service, we can just extend `JSScriptServiceUtil` and do not have to change the injection.

Comment (J-N-K) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: Much better. I have made some minor suggestions in a PR to your fork.

**Comment by jlaur** [comment]
@kaikreuzer - it seems that @jpg0 is missing as member of the addons contributors group, could you add him?

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java
Comment (florian-h05) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: @J-N-K See https://github.com/openhab/openhab-addons/pull/13695/commits/77bc18a8da5cd5615837ec561b0f3b5500e7f663 for the changes I spoke about in your PR.

**Comment by florian-h05** [comment]
FYI: DCO is failing because of @J-N-K GitHub username is not his real name, which he used to signoff.

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java
Comment (J-N-K) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: LGTM

Comment (florian-h05) [reply bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/JSScriptServiceUtil.java]: Okay, could you then please give an approving review?

**Comment by florian-h05** [comment]
@jlaur This is ready for your review and merging now.

FYI: digitaldan and jpg0 approved at some point, after that I’ve changed the implementation of `JSScriptServiceUtil` (as jpg0 and J-N-K asked me to). J-N-K gave his approval in https://github.com/openhab/openhab-addons/pull/13695#discussion_r1020812558.
After all that reviews, I only upgraded the included openhab-js version and removed some unused code.

File: bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java
Comment (jlaur) [inline bundles/org.openhab.automation.jsscripting/src/main/java/org/openhab/automation/jsscripting/internal/threading/ThreadsafeTimers.java:117]:
```suggestion
        }, identifier + ".timeout." + id, Instant.now().plusMillis(delay));
```

**Review by jlaur — CHANGES_REQUESTED** [review]
Thanks! Only a small simplification/optimization suggested.

**Comment by florian-h05** [comment]
@jlaur I addressed your review, let's wait for the CI builds and then merge.

**Review by jlaur — APPROVED** [review]
LGTM

**Comment by florian-h05** [comment]
@jlaur All checks have passed.

**Pull request merged**
